### PR TITLE
Introduce /sourcemap endpoint to allow for JS Events symbolication

### PR DESCRIPTION
**Pull request by kamilogorek**
This is very much WIP and this PR is sent only to make it easier to review and discuss the possible solutions.

Based on this, I'll try to link it together with the local Sentry instance and see whether we maybe are missing some requirements in this implementation.

Sentry PR: https://github.com/getsentry/sentry/pull/43951
Umbrella issue: https://github.com/getsentry/team-stacktrace/issues/14

#skip-changelog

### CODE REVIEW

File: crates/symbolicator-service/src/services/symbolication/sourcemap.rs [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+    urls
+}
+
+// Joins together frames `abs_path` and discovered sourcemap reference.
+fn resolve_sourcemap_url(abs_path: &str, sourcemap_location: &str) -> anyhow::Result<String> {
+    let base = Url::parse(abs_path)?;
+    base.join(sourcemap_location)
+        .map(|url| url.to_string())
+        .map_err(|e| e.into())
+}
```
Comment (loewenheim) [inline crates/symbolicator-service/src/services/symbolication/sourcemap.rs:391]: I would suggest combining these two functions into one.

1. `resolve_sourcemap_url` is only called immediately before `get_release_file_candidate_urls`.
1. The error returned by `resolve_sourcemap_url` is never handled, it just gets silently discarded by `if let Ok(…)`.
2. The functions interface via a `String` when they are really both dealing with `Url`s.

File: crates/symbolicator-service/src/services/symbolication/sourcemap.rs [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    pub sourcemap_artifact: Option<SearchArtifactResult>,
+
+    pub smcache: Option<OwnedSourceMapCache>,
+}
+
+impl SymbolicationActor {
+    async fn collect_stacktrace_artifacts(
+        &self,
+        request: &JsProcessingSymbolicateStacktraces,
+    ) -> HashMap<String, FrameSourceMapping> {
```
Comment (loewenheim) [inline crates/symbolicator-service/src/services/symbolication/sourcemap.rs:177]: As far as I can tell, the `FrameSourceMapping` struct is only used for `SourceMapCache` construction and never actually used outside of this function. This function might as well just return a map from `String` to `OwnedSourceMapCache`.
```suggestion
    async fn collect_stacktrace_artifacts(
        &self,
        request: &JsProcessingSymbolicateStacktraces,
    ) -> HashMap<String, OwnedSourceMapCache> {
```

**Review by loewenheim — COMMENTED** [review]
Very impressive work!

**Comment by kamilogorek** [comment]
@loewenheim all naming is up for discussion, and anything called `js_processing_` or `JsProcessingX` is just a placeholder for now :)

File: crates/symbolicator-service/src/services/download/mod.rs [hunk 3]
```diff
@@ -248,6 +250,31 @@ impl DownloadService {
         }
         remote_files
     }
+
+    pub async fn list_artifacts(
+        &self,
+        source: Arc<SentrySourceConfig>,
+    ) -> Vec<SearchArtifactResult> {
```
[the remark this answers is not in the document]
Comment (kamilogorek) [reply crates/symbolicator-service/src/services/download/mod.rs:257]: I doubt that tbh, as it's heavily dependent on the internals of our own Sentry API, and requires calls to its endpoints. I don't see anyone ever replicate it 😅

File: crates/symbolicator-service/src/types/mod.rs [hunk 4]
```diff
@@ -530,6 +552,18 @@ pub struct CompletedSymbolicationResponse {
...
 }
 
+#[derive(Debug, Default, Clone, Deserialize, Serialize)]
+pub struct JsProcessingCompletedSymbolicationResponse {
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        with = "chrono::serde::ts_seconds_option"
+    )]
+    pub timestamp: Option<DateTime<Utc>>,
```
Comment (Swatinem) [inline crates/symbolicator-service/src/types/mod.rs:562]: do we even need this?

File: crates/symbolicator-service/src/services/symbolication/sourcemap.rs [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+
+            for candidate in get_release_file_candidate_urls(&abs_path) {
+                if let Some(source_artifact) = release_archive.get(&candidate) {
+                    if let Some(source_file) = self
+                        .sourcemaps
+                        .fetch_artifact(SentryRemoteFile::new(
+                            request.source.clone(),
+                            source_artifact.id.clone(),
+                            SentryFileType::ReleaseArtifact,
+                        ))
```
[the remark this answers is not in the document]
Comment (Swatinem) [reply crates/symbolicator-service/src/services/symbolication/sourcemap.rs:201]: That is a very good idea actually. Might move the problem of `SentryFileType` a layer inwards.

File: crates/symbolicator-service/src/services/symbolication/sourcemap.rs [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+                        .await
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/symbolication/sourcemap.rs:202]: this await is "blocking" the loop here, same for the sourcemap reference below.
Ideally we would `map` the set of `unique_abs_paths` into a list of futures for fetching source files concurrently.

but we can also leave this for later, as this is the core of the caching layer we have to figure out.

File: crates/symbolicator-service/src/services/symbolication/sourcemap.rs [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+    stacktrace: JsProcessingRawStacktrace,
+    artifacts: &HashMap<String, FrameSourceMapping>,
+) -> CompleteJsStacktrace {
+    let mut symbolicated_frames = vec![];
+    let unsymbolicated_frames_iter = stacktrace.frames.into_iter();
+
+    for mut frame in unsymbolicated_frames_iter {
+        match js_processing_symbolicate_frame(&mut frame, &artifacts) {
+            // TODO: Should frames filtering (eg. unsupported platform, or non-in_app)
+            // be done here or by monolith?
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/symbolication/sourcemap.rs:278]: for native, it happens in the monolith and I would suggest to do that the same.

File: crates/symbolicator-service/src/services/symbolication/sourcemap.rs [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+        if let Some(smcache) = &artifact.smcache {
+            let token = smcache
+                .get()
+                .lookup(SourcePosition::new(
+                    frame.lineno.unwrap() - 1,
+                    frame.colno.unwrap() - 1,
+                ))
+                .unwrap();
+
+            let file = token.file().unwrap();
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/symbolication/sourcemap.rs:308]: there is a lot of `unwrap` all over the place :-) We should decide how to do error handling here.
What does the current code do if some source was not fetched successfully? I believe it is reporting these to the user somehow?

File: crates/symbolicator-service/src/services/symbolication/sourcemap.rs [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+    pub source_artifact: Option<SearchArtifactResult>,
+
+    pub sourcemap_file: Option<NamedTempFile>,
+    pub sourcemap_artifact: Option<SearchArtifactResult>,
+
+    pub smcache: Option<OwnedSourceMapCache>,
+}
+
+impl SymbolicationActor {
+    async fn collect_stacktrace_artifacts(
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/symbolication/sourcemap.rs:174]: I would move this and related functions to its own file, similar to how `ModuleLookup` works for native. So you have fetching / transforming of the files in one place, and using them in another.

**Review by Swatinem — COMMENTED** [review]
looks like a very good start so far.

We should have a look at how the monolith code deals with error handling, and we should expose any errors that happen in the result, similar to how we do that for candidates right now.

**Comment by kamilogorek** [comment]
- most code review commented resolved
- rebased on top of master
- cleaned up most gibberish
- commented with `TODO(sourcemap)` where appropriate

**Review by Swatinem — APPROVED** [review]

**Comment by kamilogorek** [comment]
wat

**Pull request merged**